mindmeld/components/question_answerer.py

```diff
@@ -449,805 +591,2172 @@ def _generate_mapping_data(embedder_model, embedding_fields):
+
+        @staticmethod
+        def is_location(value):
+            return (
+                (isinstance(value, dict) and "lat" in value and "lon" in value)
+                or (isinstance(value, list) and len(value) == 2 and
+                    isinstance(value[0], numbers.Number) and isinstance(value[1], numbers.Number))
+                or (isinstance(value, str) and "," in value and len(value.split(",")) == 2)
             )
-            raise KnowledgeBaseError from e
-        except ElasticsearchException as e:
-            raise KnowledgeBaseError from e
 
-    class Clause(ABC):
-        """This class models an abstract knowledge base clause."""
+        @staticmethod
+        def number_scorer(some_number):
+            return some_number
 
-        def __init__(self):
-            """Initialize a knowledge base clause"""
-            self.clause_type = None
+        @staticmethod
+        def date_scorer(some_date, origin_date=datetime.datetime.now()):
+            """
+            ascertains a suitable date format for input and
+            returns number of days from origin date as score
+            """
+            target_date = origin_date
+            for fmt in NonElasticsearchQuestionAnswerer.FieldResource.DATE_FORMATS:
+                try:
+                    target_date = datetime.datetime.strptime(some_date, fmt)
+                    break
+                except ValueError:
+                    pass
+            return (target_date - origin_date).days
+
+        @staticmethod
+        def location_scorer(some_location, source_location):
+            """
+            Uses Haversine formula to find distance between two coordinates
+            references: https://en.wikipedia.org/wiki/Haversine_formula and
+                        http://www.movable-type.co.uk/scripts/latlong.html
+
+            Args:
+                some_location (str): latitude and longitude supplied as
+                    comma separated strings, eg. "37.77,122.41"
+                source_location (str): latitude and longitude supplied as
+                    comma separated strings, eg. "37.77,122.41"
+
+            Returns:
+                float: distance between the coordinates in kilometers
+
+            Example 1:
+                >>> point_1 = "37.78953146306901,-122.41160227491551" # SF in CA
+                >>> point_2 = "47.65182346406002, -122.36765696283909" # Seattle in WA
+                >>> location_scorer(point_1, point_2)
+                >>> # 1096.98 kms (approx. points on Google maps says 680.23 miles/1094.72 kms)
+            Example 2:
+                >>> point_3, point_4 = "52.2296756,21.0122287", "52.406374,16.9251681"
+                >>> location_scorer(point_3, point_4)
+                >>> # 278.54 kms
+            """
 
-        @abstractmethod
-        def validate(self):
-            """Validate the clause."""
-            raise NotImplementedError("Must override validate()")
+            R = 6373.0
```

Comment:
Whats the significance of this constant?

Reply:
it is just a constant based on the _Haversine formula_